From: Patrick Brunschwig <patrick@enigmail.net>
Date: Wed, 2 Nov 2016 17:34:43 +0100
Subject: fixed race condition for short lived processes,
 potentially returning the wrong exit code. Patch by Kai Michaelis

---
 ipc/modules/subprocess.jsm            | 22 ++++++++++------------
 ipc/modules/subprocess_worker_unix.js |  4 ++--
 2 files changed, 12 insertions(+), 14 deletions(-)

diff --git a/ipc/modules/subprocess.jsm b/ipc/modules/subprocess.jsm
index d0ae392..e465ba0 100644
--- a/ipc/modules/subprocess.jsm
+++ b/ipc/modules/subprocess.jsm
@@ -1123,7 +1123,6 @@ function subprocess_unix(options) {
     active = true,
     done = false,
     exitCode = -1,
-    workerExitCode = 0,
     child = {},
     pid = -1,
     writeWorker = [],
@@ -1657,7 +1656,7 @@ function subprocess_unix(options) {
           break;
         case "done":
           debugLog("Pipe " + name + " closed\n");
-          if (event.data.data !== 0) workerExitCode = event.data.data;
+          updateExitCode(event.data.data);
           --readers;
           if (readers === 0) cleanup();
           break;
@@ -1737,6 +1736,13 @@ function subprocess_unix(options) {
     }
   }
 
+  function updateExitCode(code) {
+    if (exitCode > -2 && code >= 0) {
+      exitCode = code;
+    }
+    exitCode = exitCode % 0xFF;
+  }
+
   function cleanup() {
     debugLog("Cleanup called\n");
 
@@ -1753,18 +1759,10 @@ function subprocess_unix(options) {
       var result, status = ctypes.int();
       result = waitpid(child.pid, status.address(), 0);
 
-      if (exitCode > -2) {
-        if (result > 0)
-          exitCode = status.value;
-        else
-        if (workerExitCode >= 0)
-          exitCode = workerExitCode;
-        else
-          exitCode = status.value;
+      if (exitCode > -2 && result > 0) {
+        updateExitCode((status.value & 0xff00) >> 8);
       }
 
-      exitCode = exitCode % 0xFF;
-
       for (i = 0; i < writeWorker.length; i++) {
         if (writeWorker[i])
           writeWorker[i].postMessage({
diff --git a/ipc/modules/subprocess_worker_unix.js b/ipc/modules/subprocess_worker_unix.js
index ca4507b..03a0fe1 100644
--- a/ipc/modules/subprocess_worker_unix.js
+++ b/ipc/modules/subprocess_worker_unix.js
@@ -252,10 +252,10 @@ function readPipe(pipe, charset, pid, bufferedOutput) {
       result = libcFunc.waitpid(pid, status.address(), WNOHANG);
       if (result > 0) {
         pollTimeout = NOWAIT;
-        exitCode = parseInt(status.value, 10);
+        exitCode = (parseInt(status.value, 10) & 0xff00) >> 8;
         postMessage({
           msg: "debug",
-          data: "waitpid signaled subprocess stop, exitcode=" + status.value
+          data: "waitpid signaled subprocess stop, exitcode=" + exitCode
         });
       }
       else if (result < 0) {
